### PR TITLE
[#11756] Session closing email contains no information about extended deadline

### LINKED ISSUE #11756: Session closing email contains no information about extended deadline
- **Environment**: V8.14.1

**Steps to reproduce**

1. Create a session
1. Grant deadline extensions to some students/instructors
1. Wait for these students/instructors to obtain their "session is closing soon" email

**Expected behaviour**

The "session is closing soon" email contains information about the extended deadline.

**Actual behaviour**

The "session is closing soon" email contains information about the original deadline, and there is no mention about the extended deadline in any form. This can and will be a major source of confusion.

### PR DESCRIPTION
**Pull request by samuelfangjw**
Fixes #11756

**Outline of Solution**

- Update deadline to show correct deadline after extension is accounted for
- If deadline has been extended, update deadline to `... (after extension)`

Screen Shot 2022-04-22 at 10 51 18 PM

### CODE REVIEW

**Comment by samuelfangjw** [comment]
> Instead of `Deadline (Extended): ...`, why not `Deadline: ... (after extension)`? This should significantly simplify some implementation while not being any less clear.

Sure, thanks for the suggestions.

**Review by FergusMok — APPROVED** [review]
LGTM, thanks for the quick fix.

On another note, would it be good to have in the future for a reminder when extensions are ending? (Basically what this email does for non-extendees).

**Comment by damithc** [comment]
> On another note, would it be good to have in the future for a reminder when extensions are ending? (Basically what this email does for non-extendees).

Good point. As this email is sent to the whole class, I guess probably no need for additional reminders.

**Comment by samuelfangjw** [comment]
> On another note, would it be good to have in the future for a reminder when extensions are ending? (Basically what this email does for non-extendees).

This email is sent about a day before the extension ends. In the general case, the student receives an email about the extension when it is granted, followed by this email just before his extension is over (following the new definition of session closed defined in the tech design).

File: src/main/java/teammates/logic/api/EmailGenerator.java [hunk 1]
```diff
@@ -521,8 +520,29 @@ public List<EmailWrapper> generateFeedbackSessionClosingWithExtensionEmails(
...
+        for (StudentAttributes student : students) {
+            FeedbackSessionAttributes studentSession = session.getCopyForStudent(student.getEmail());
+            emails.add(generateFeedbackSessionEmailBaseForStudents(course, studentSession, student,
+                    template, type, feedbackAction, studentAdditionalContactBuilder.toString()));
+        }
+        for (InstructorAttributes instructor : instructors) {
+            FeedbackSessionAttributes instructorSession = session.getCopyForInstructor(instructor.getEmail());
+            emails.add(generateFeedbackSessionEmailBaseForInstructors(course, instructorSession, instructor,
+                    template, type, feedbackAction, instructorAdditionalContactBuilder.toString()));
+        }
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/api/EmailGenerator.java:544]: Does this work:
```java
for (StudentAttributes student : students) {
    FeedbackSessionAttributes studentSession = session.getCopyForStudent(student.getEmail());
    emails.add(generateFeedbackSessionEmailBases(course, studentSession, Collections.singletonList(student),
            new ArrayList<>(), new ArrayList<>(), template, type, feedbackAction));
}
for (InstructorAttributes instructor : instructors) {
    FeedbackSessionAttributes instructorSession = session.getCopyForInstructor(instructor.getEmail());
    emails.add(generateFeedbackSessionEmailBases(course, instructorSession, new ArrayList<>(),
            Collections.singletonList(instructor), new ArrayList<>(), template, type, feedbackAction));
}
```
i.e. no need to repeat the additional contact builders

**Review by wkurniawan07 — COMMENTED** [review]
Ok to accept as is, but a minor suggestion

File: src/main/java/teammates/logic/api/EmailGenerator.java (hunk 1 above)
Comment (samuelfangjw) [reply src/main/java/teammates/logic/api/EmailGenerator.java:544]: Yes this will work too. Will make the change.

Comment (samuelfangjw) [reply src/main/java/teammates/logic/api/EmailGenerator.java:536]: Changed to the following based on your suggestion:
https://github.com/TEAMMATES/teammates/blob/d74d1500092a6ae21fe670016cbb3478de2a0d0b/src/main/java/teammates/logic/api/EmailGenerator.java#L527-L536

**Review by wkurniawan07 — APPROVED** [review]

**Pull request merged**